File: 000Unordered_ Videos/WWDC 2013/WWDC 2013 Sample Code/Folder/TicTacToeApp/TicTacToe/TTTMove.m
/*
     File: TTTMove.m
 Abstract: 
 
  Version: 1.0
 
 Disclaimer: IMPORTANT:  This Apple software is supplied to you by Apple
 Inc. ("Apple") in consideration of your agreement to the following
 terms, and your use, installation, modification or redistribution of
 this Apple software constitutes acceptance of these terms.  If you do
 not agree with these terms, please do not use, install, modify or
 redistribute this Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a personal, non-exclusive
 license, under Apple's copyrights in this original Apple software (the
 "Apple Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple.  Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis.  APPLE
 MAKES NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION
 THE IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS
 FOR A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 Copyright (C) 2013 Apple Inc. All Rights Reserved.
 
 
 Copyright © 2013 Apple Inc. All rights reserved.
 WWDC 2013 License
 
 NOTE: This Apple Software was supplied by Apple as part of a WWDC 2013
 Session. Please refer to the applicable WWDC 2013 Session for further
 information.
 
 IMPORTANT: This Apple software is supplied to you by Apple Inc.
 ("Apple") in consideration of your agreement to the following terms, and
 your use, installation, modification or redistribution of this Apple
 software constitutes acceptance of these terms. If you do not agree with
 these terms, please do not use, install, modify or redistribute this
 Apple software.
 
 In consideration of your agreement to abide by the following terms, and
 subject to these terms, Apple grants you a non-exclusive license, under
 Apple's copyrights in this original Apple software (the "Apple
 Software"), to use, reproduce, modify and redistribute the Apple
 Software, with or without modifications, in source and/or binary forms;
 provided that if you redistribute the Apple Software in its entirety and
 without modifications, you must retain this notice and the following
 text and disclaimers in all such redistributions of the Apple Software.
 Neither the name, trademarks, service marks or logos of Apple Inc. may
 be used to endorse or promote products derived from the Apple Software
 without specific prior written permission from Apple. Except as
 expressly stated in this notice, no other rights or licenses, express or
 implied, are granted by Apple herein, including but not limited to any
 patent rights that may be infringed by your derivative works or by other
 works in which the Apple Software may be incorporated.
 
 The Apple Software is provided by Apple on an "AS IS" basis. APPLE MAKES
 NO WARRANTIES, EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION THE
 IMPLIED WARRANTIES OF NON-INFRINGEMENT, MERCHANTABILITY AND FITNESS FOR
 A PARTICULAR PURPOSE, REGARDING THE APPLE SOFTWARE OR ITS USE AND
 OPERATION ALONE OR IN COMBINATION WITH YOUR PRODUCTS.
 
 IN NO EVENT SHALL APPLE BE LIABLE FOR ANY SPECIAL, INDIRECT, INCIDENTAL
 OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 INTERRUPTION) ARISING IN ANY WAY OUT OF THE USE, REPRODUCTION,
 MODIFICATION AND/OR DISTRIBUTION OF THE APPLE SOFTWARE, HOWEVER CAUSED
 AND WHETHER UNDER THEORY OF CONTRACT, TORT (INCLUDING NEGLIGENCE),
 STRICT LIABILITY OR OTHERWISE, EVEN IF APPLE HAS BEEN ADVISED OF THE
 POSSIBILITY OF SUCH DAMAGE.
 
 EA1002
 5/3/2013
 */

#import "TTTMove.h"

NSInteger const TTTMoveSidePositionsCount = 3;

@implementation TTTMove

- (instancetype)initWithPlayer:(TTTMovePlayer)player xPosition:(TTTMoveXPosition)xPosition yPosition:(TTTMoveYPosition)yPosition
{
    self = [super init];
    if (self) {
        _player = player;
        _xPosition = xPosition;
        _yPosition = yPosition;
    }
    return self;
}

- (id)init
{
    self = [self initWithPlayer:TTTMovePlayerMe xPosition:TTTMoveXPositionCenter yPosition:TTTMoveYPositionCenter];
    return self;
}

static NSString * const TTTMoveEncodingKeyPlayer = @"player";
static NSString * const TTTMoveEncodingKeyXPosition = @"xPosition";
static NSString * const TTTMoveEncodingKeyYPosition = @"yPosition";

- (id)initWithCoder:(NSCoder *)coder
{
    self = [super init];
    if (self) {
        _player = [coder decodeIntegerForKey:TTTMoveEncodingKeyPlayer];
        _xPosition = [coder decodeIntegerForKey:TTTMoveEncodingKeyXPosition];
        _yPosition = [coder decodeIntegerForKey:TTTMoveEncodingKeyYPosition];
    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)coder
{
    [coder encodeInteger:self.player forKey:TTTMoveEncodingKeyPlayer];
    [coder encodeInteger:self.xPosition forKey:TTTMoveEncodingKeyXPosition];
    [coder encodeInteger:self.yPosition forKey:TTTMoveEncodingKeyYPosition];
}

- (NSUInteger)hash
{
    NSUInteger hash = 1;
    hash = 31 * hash + self.player;
    hash = 31 * hash + self.xPosition + 1;
    hash = 31 * hash + self.yPosition + 1;
    return hash;
}

- (BOOL)isEqual:(id)object
{
    if (![object isKindOfClass:[TTTMove class]]) {
        return NO;
    }
    
    TTTMove *move = object;
    return ((self.player == move.player) && (self.xPosition == move.xPosition) && (self.yPosition == move.yPosition));
}

@end

File: UIKit and Animations/WWDC 2013/Customizing Your App's Appearance for iOS 7/Sample_Code/TicTacToeApp_Swift/TicTacToeApp_Swift/Model/TTTMove.swift
//
//  TTTMove.swift
//  TicTacToeApp_Swift
//
//  Created by Seyed Samad Gholamzadeh on 7/9/18.
//  Copyright © 2018 Seyed Samad Gholamzadeh. All rights reserved.
//

import UIKit

enum TTTMovePlayer: Int {
	case me, enemy
}

enum TTTMoveXPosition: Int {
	case left = -1
	case center = 0
	case right = 1
}

enum TTTMoveYPosition: Int {
	case top = -1
	case center = 0
	case bottom = 1
}

class TTTMove: NSObject, NSCoding {

	let TTTMoveEncodingKeyPlayer = "player"
	let TTTMoveEncodingKeyXPosition = "xPosition"
	let TTTMoveEncodingKeyYPosition = "yPosition"

	var player: TTTMovePlayer!
	var xPosition: TTTMoveXPosition!
	var yPosition: TTTMoveYPosition!
	
	init(player: TTTMovePlayer, xPosition: TTTMoveXPosition, yPosition: TTTMoveYPosition) {
		super.init()
		self.player = player
		self.xPosition = xPosition
		self.yPosition = yPosition
	}
	
	override convenience init() {
		self.init(player: .me, xPosition: .center, yPosition: .center)
	}
	
	required init?(coder aDecoder: NSCoder) {
		super.init()
		self.player = TTTMovePlayer(rawValue: aDecoder.decodeInteger(forKey: TTTMoveEncodingKeyPlayer))
		self.xPosition = TTTMoveXPosition(rawValue: aDecoder.decodeInteger(forKey: TTTMoveEncodingKeyXPosition))
		self.yPosition = TTTMoveYPosition(rawValue: aDecoder.decodeInteger(forKey: TTTMoveEncodingKeyYPosition))

	}
	
	func encode(with aCoder: NSCoder) {
		aCoder.encode(self.player.rawValue, forKey: TTTMoveEncodingKeyPlayer)
		aCoder.encode(self.xPosition.rawValue, forKey: TTTMoveEncodingKeyXPosition)
		aCoder.encode(self.yPosition.rawValue, forKey: TTTMoveEncodingKeyYPosition)
	}
	
	override var hash: Int {
		var hash: Int = 1
		hash = 31 * hash + self.player.rawValue
		hash = 31 * hash + self.xPosition.rawValue + 1
		hash = 31 * hash + self.yPosition.rawValue + 1
		return hash
	}
	
	override func isEqual(_ object: Any?) -> Bool {
		if !(object is TTTMove) {
			return false
		}
		
		let move = object as! TTTMove
		return self.player == move.player && self.xPosition == move.xPosition && self.yPosition == move.yPosition
	}
	
}
